### PR TITLE
PresenceEventEngine integration

### PR DESCRIPTION
**Pull request by mohitpubnub**
Presence Event Engine integration with existing api
Handing setState when EnableEventEngine is true
Manage Unsubscribe with EventEngine

### CODE REVIEW

**Comment by mohitpubnub** [comment]
@Xavrax Ready for review

File: src/Api/PubnubApi/EndPoint/PubSub/UnsubscribeOperation.cs [hunk 1]
```diff
@@ -54,7 +58,30 @@ public UnsubscribeOperation<T> QueryParam(Dictionary<string, object> customQuery
...
+            if (config.EnableEventEngine && subscribeEventEngineFactory != null) {
+                if (subscribeEventEngineFactory.hasEventEngine(PubnubInstance.InstanceId)) {
+                    var subscribeEventEngine = subscribeEventEngineFactory.getEventEngine(PubnubInstance.InstanceId);
+                    subscribeEventEngine.Channels = subscribeEventEngine.Channels.Except(this.subscribeChannelNames).ToArray();
+                    subscribeEventEngine.Channelgroups = subscribeEventEngine.Channelgroups.Except(this.subscribeChannelGroupNames).ToArray();
+                    subscribeEventEngine.eventQueue.Enqueue(new SubscriptionChangedEvent() {
+                        Channels = subscribeEventEngine.Channels, ChannelGroups = subscribeEventEngine.Channelgroups
+                    });
+                    if (ChannelLocalUserState.TryGetValue(PubnubInstance.InstanceId, out var userState)) {
+                        foreach (var channel in this.subscribeChannelNames ?? Array.Empty<string>()) {
```
Comment (Xavrax) [inline src/Api/PubnubApi/EndPoint/PubSub/UnsubscribeOperation.cs:70]: This line doesn't compile for some versions of dotnet. 
```suggestion
                        foreach (var channel in this.subscribeChannelNames ?? new string[0]) {
```

File: src/Api/PubnubApi/EndPoint/PubSub/UnsubscribeOperation.cs [hunk 2]
```diff
@@ -54,7 +58,30 @@ public UnsubscribeOperation<T> QueryParam(Dictionary<string, object> customQuery
...
+                            userState.TryRemove(channel, out _);
+                        }
+                    }
+                    if (ChannelGroupLocalUserState.TryGetValue(PubnubInstance.InstanceId, out var channelGroupUserState)) {
+                        foreach (var channelGroup in this.subscribeChannelGroupNames ?? Array.Empty<string>()) {
```
Comment (Xavrax) [inline src/Api/PubnubApi/EndPoint/PubSub/UnsubscribeOperation.cs:75]:
```suggestion
                        foreach (var channelGroup in this.subscribeChannelGroupNames ?? new string[0]) {
```

**Review by Xavrax — APPROVED** [review]
LGTM - but we discussed - EE is not working at all. :/

File: src/Api/PubnubApi/EndPoint/PubSub/UnsubscribeOperation.cs (hunk 1 above)
Comment (mohitpubnub) [reply src/Api/PubnubApi/EndPoint/PubSub/UnsubscribeOperation.cs]: 👍🏻

File: src/Api/PubnubApi/EndPoint/PubSub/UnsubscribeOperation.cs (hunk 2 above)
Comment (mohitpubnub) [reply src/Api/PubnubApi/EndPoint/PubSub/UnsubscribeOperation.cs]: 👍🏻

**Pull request merged**